src/core/lib/surface/init.cc

```diff
@@ -149,37 +147,44 @@ void grpc_init(void) {
     grpc_tracer_init("GRPC_TRACE");
     /* no more changes to channel init pipelines */
     grpc_channel_init_finalize();
-    grpc_iomgr_start(&exec_ctx);
+    grpc_iomgr_start();
   }
   gpr_mu_unlock(&g_init_mu);
-  grpc_exec_ctx_finish(&exec_ctx);
+
   GRPC_API_TRACE("grpc_init(void)", 0, ());
 }
 
 void grpc_shutdown(void) {
   int i;
   GRPC_API_TRACE("grpc_shutdown(void)", 0, ());
-  grpc_exec_ctx exec_ctx =
-      GRPC_EXEC_CTX_INITIALIZER(0, grpc_never_ready_to_finish, nullptr);
+  if (grpc_core::ExecCtx::Get()) {
+    grpc_core::ExecCtx::Get()->Flush();
+  }
   gpr_mu_lock(&g_init_mu);
   if (--g_initializations == 0) {
-    grpc_executor_shutdown(&exec_ctx);
-    grpc_timer_manager_set_threading(false);  // shutdown timer_manager thread
-    for (i = g_number_of_plugins; i >= 0; i--) {
-      if (g_all_of_the_plugins[i].destroy != nullptr) {
-        g_all_of_the_plugins[i].destroy();
+    {
+      grpc_core::ExecCtx exec_ctx(0);
+      {
+        grpc_executor_shutdown();
+        grpc_timer_manager_set_threading(
+            false);  // shutdown timer_manager thread
+        for (i = g_number_of_plugins; i >= 0; i--) {
+          if (g_all_of_the_plugins[i].destroy != nullptr) {
+            g_all_of_the_plugins[i].destroy();
+          }
+        }
       }
+      gpr_timers_global_destroy();
+      grpc_tracer_shutdown();
```

Comment:
Seems risky to move before iomgr shutdown: iomgr certainly uses tracers

Reply:
That's blank, so wouldn't have affected any tests but something could have gone wrong in the future. Moved it back.That's blank, so wouldn't have affected any tests but something could have gone wrong in the future. Moved it back.